chain/chain/src/validate.rs

```diff
@@ -65,29 +68,28 @@ pub fn validate_chunk_proofs(chunk: &ShardChunk, runtime_adapter: &dyn RuntimeAd
     let (tx_root, _) = merklize(transactions);
     if tx_root != chunk.tx_root() {
         byzantine_assert!(false);
-        return false;
+        return Ok(false);
     }
     // 2c. Checking that chunk receipts are valid
     if height_created == 0 {
-        return receipts.len() == 0 && outgoing_receipts_root == CryptoHash::default();
+        return Ok(receipts.len() == 0 && outgoing_receipts_root == CryptoHash::default());
     } else {
-        let epoch_id = {
+        let shard_layout = {
             let prev_block_hash = match chunk {
                 ShardChunk::V1(chunk) => chunk.header.inner.prev_block_hash,
                 ShardChunk::V2(chunk) => chunk.header.prev_block_hash(),
             };
-            runtime_adapter.get_epoch_id_from_prev_block(&prev_block_hash).unwrap()
```

Comment:
Not really relevant to this PR, but I spot a few unwraps here added by me before and decided to correct them here

Reply:
I am actually thinking if perhaps such unwraps are fine? 

That is, I *think* our general strategy is "node should shut-down itself in the face of catastrophic failures". If that's indeed the case, I see an argument for initiating shutdown at the place where the error originates, not at the top-level where we handle the errors. 

There's a certain beauty if, in the API level, the database is reliable and never returns an `io::Error`. 


Though, this is utterly unrelate to the PR, and we already generally just `?` errors

Reply:
I think the struggle here is that it's hard to distinguish from within the function whether the error is catastrophic. I always struggled to find a good way to deal with this, and I am glad you brought this up. 

Focusing on this specific case, this error will occur if by the time this function is called, `prev_block_hash` is not processed yet. This almost should never happen because we would probably only validate chunks after the previous block is processed. But it's hard to guarantee that. Unwrapping here seems risky because all callers of this function then need to know that this function may panic in the case of `prev_block` not processed. What if someone uses this function incorrectly in the future, or maybe simply I just understand this function wrong. On the other hand, passing along the the error also doesn't seem ideal because it could be a critical failure (maybe the database is corrupt) and it is easy for errors to get lost and not handled correctly if we keep propagating them.

Reply:
For general applications, I really love sqlite approach for this(https://www.sqlite.org/assert.html): crash in debug, in release log error and return a "neutral" value (ie, error object). 

For blockchain, not really sure whats the best apporach. 

OTOH, sqlite is aviation grade software, so if thinks it ok to return newtral value, we can do the same.

:thinking: I've just realized that a lot of tradeoffs we face feel similar to those of sqlite, and that is a brilliant piece of engineering. I need to re-read its docs to figure out what we might want to apply in nearcore.

Reply:
Yeah I think that's a great point! With better test coverage, we can afford to be more assertive in error handling. 

At my previous company (a trading firm) we did that too since we don't want the trading program just crash in live but we also don't want miss any catastrophic error and lose a lot of money in live. One thing we did was aggressively assert, which panic in debug but log error in release. We then run lots of simulations (regression tests) using the debug binary to make sure nothing crashes. In live we would log the errors and ops will get sound notifications (something like "oh no we are all doomed" lol) every time when an error is logged. 

Maybe we can apply that to betanet vs testnet/mainnet. In betanet we use the debug binary which will fail aggressively in case of error. 

I think we should also have more "simulation" kind of tests in near which can use real historic data and cover more complicated scenarios (or maybe we already have load test that is similar to that. I am not very familiar with our test framework). But that's more complicated .